It is bad solution, and program doesn't work. I will try to fix it ne…

Y-Pazniak opened the pull request:
…xt week

Y-Pazniak committed:
It is bad solution, and program doesn't work. I will try to fix it next week

Y-Pazniak committed:
It is bad solution, and program doesn't work. I will try to fix it next week

MaximPilipeiko reviewed (changes requested):
Please, ping me, when it will be ready to review

Y-Pazniak committed:
Operations with UserProgress were removed from the controller (MessageRouter) to the service (UserProgressManager)

Y-Pazniak committed:
Operations with UserProgress were removed from the controller (MessageRouter) to the service (UserProgressManager)

Y-Pazniak committed:
It seems like I have managed to add saving in file and it works properly. It is not the best way, I guess, but at least it works.

Y-Pazniak committed:
New user creation added after "/start" command.

MaximPilipeiko commented on pom.xml:
```diff
@@ -38,8 +38,8 @@
                 <configuration>
-                    <source>21</source>
-                    <target>21</target>
+                    <source>19</source>
```
Why downgrade?

Y-Pazniak replied:
Sorry, my mistake. The case of error is my inattention.
Fixed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -36,22 +36,24 @@ public SendMessage route(Update update) {
+            }
             case Command.EAES, Command.OTHER_COUNTRIES ->
-                    answer = botFieldsManager.processCarOrigin(usersMessage, userProgress);
+                    answer = userProgressManager.processCarOrigin(usersMessage, message);
```
just userId should be enough.
If it's possible it's better that `userProgressManager` don't know about telegram api.
```suggestion
                    answer = userProgressManager.processCarOrigin(userId, usersMessage);
```

Y-Pazniak replied:
Fixed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgress.java:
```diff
@@ -2,19 +2,18 @@
     private CarAge carAge = null;
     private TypeOfEngine typeOfEngine = null;
     private VolumeOfEngine volumeOfEngine = null;
+    private final String chatID;
```
Please, for consistency use something one `chatId` or `userId`

Y-Pazniak replied:
chatId  is better. I renamed all the variables into the same way

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgressStorage.java:
```diff
@@ -1,27 +1,52 @@
+        private static final UserProgressStorage USER_PROGRESS_STORAGE = new UserProgressStorage();
+    }
+
+    private HashMap<String, UserProgress> users;
```
Please, use abstraction instead implementation
```suggestion
    private Map<String, UserProgress> users;
```

Y-Pazniak replied:
Sorry, bad habit. Fixed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgressStorage.java:
```diff
@@ -1,27 +1,52 @@
+import java.io.*;
 import java.util.HashMap;
 
 public class UserProgressStorage {
```
Please, create interface for UserProgressStorage. Could be done in future. Lets discuss it during call.
```suggestion
public class UserProgressStorage {
```

Y-Pazniak replied:
Realized. Looks great

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgressStorage.java:
```diff
@@ -1,27 +1,52 @@
+        private static final UserProgressStorage USER_PROGRESS_STORAGE = new UserProgressStorage();
+    }
+
+    private HashMap<String, UserProgress> users;
```
I recommend removing `Map` and storing data only in file(s). I remember that you liked idea of storing every user in separate file. Maybe you could implement it?

Y-Pazniak replied:
Realized. It was easyer than I thought. I am not sure about the code, but at least it seems to work. I am happy. Thanks for yout support!

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -61,12 +76,15 @@ public String processEngineVolume(String command, UserProgress userProgress) {
     }
 
-
+    private UserProgress getUserProgress(Message message) {
```
I think this metod is redundant

Y-Pazniak replied:
Deleted.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/UserProgressStorage.java:
```diff
@@ -1,27 +1,52 @@
 
-    private static class UserProgressStorageHolder {
-        private static final UserProgressStorage USER_PROGRESS_STORAGE = new UserProgressStorage();
+    public void save() {
```
It's strange that `save` do not accept any params. What does it save? It's not clear.
```suggestion
    public void save(UserProgress userProgress) {
```

Y-Pazniak replied:
Fixed.

Y-Pazniak committed:
I have added the interface. Also separate files saving has been added. I have tested it on several phones, it seems it works fine.
I am not sure about optimal way to work with files, but at least it works.

Y-Pazniak committed:
Exception UserNotFound added.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -35,28 +35,30 @@ public SendMessage route(Update update) {
             default -> answer = getSorryMessage();
         }
 
         SendMessage sendMessage = new SendMessage();
```
I recommend to use appropriate constructor and remove default(without params) constructor. In this case, it's impossible to create "invalid" object, an object with `null` fields. 
```suggestion
        SendMessage sendMessage = new SendMessage(chatId, answer);
```

Y-Pazniak replied:
The brilliant idea, i like it. Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+import java.nio.file.Paths;
+
+public class FileUserProgressStorage implements iUserProgressStorage {
+    private final String path = "D:/%s.dat";
```
Please, move `"D:/"` to env variable. In this case you can run your code on machine where there is no disk "D:/" without code changes.

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class FileUserProgressStorage implements iUserProgressStorage {
```
```suggestion
public class FileUserProgressStorage implements  IUserProgressStorage {
```

Y-Pazniak replied:
Renamed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public void save(UserProgress userProgress) {
+        try {
+            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(String.format(path, userProgress.getChatID())));
```
What do you think to encapsulate building path to file logic in dedicated method?
```suggestion
            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(buildFilePath(path, userProgress.getChatID())));
```

Y-Pazniak replied:
Good point. Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public UserProgress get(String chatID) throws UserFileNotFoundException {
+        UserProgress localUserProgress;
+        if (isUserFileExists(chatID)) {
```
To avoid redundant inner block there is "fail-fast" approach:
```suggestion
        if (!isUserFileExists(chatID)) {
            throw new UserFileNotFoundException();
        }
        try {
                ObjectInputStream ois = new ObjectInputStream(new FileInputStream(String.format(path, chatID)));
               return (UserProgress) ois.readObject();
            } catch (IOException | ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
```

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+                localUserProgress = (UserProgress) ois.readObject();
+                return localUserProgress;
+            } catch (IOException | ClassNotFoundException e) {
+                throw new RuntimeException(e);
```
I recommend using a custom exception. `ReadingUserProgressFromFileException` or smth like this.
And let's try to avoid RuntimeException (it's not a strict rule).

Y-Pazniak replied:
New excertion was added.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public UserProgress get(String chatID) throws UserFileNotFoundException {
```
And for other methods please use final with params.
```suggestion
    public UserProgress get(final String chatID) throws UserFileNotFoundException {
```

Y-Pazniak replied:
OK. Let's discuss, why.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/MapUserProgressStorage.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void save(UserProgress userProgress) {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream("D:/users.dat"))) {
```
I think, store in file is redundant

Y-Pazniak replied:
Haven't done it yet. 
Are you sure we need to fix MapUserProgressStorage? I used this file to train skill of saving into file. We can try to make smth meaningful from this class, but it is like an outdated draft, in my honest opinion.

MaximPilipeiko replied:
Let's discuss it on call. If it's just for the train, it's ok. But I prefer to keep the codebase as pure as possible.

Y-Pazniak replied:
Fixed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/iUserProgressStorage.java:
```diff
@@ -0,0 +1,9 @@
+
+import by.custom.utilcalculator.exception.UserFileNotFoundException;
+
+public interface iUserProgressStorage {
```
```suggestion
public interface IUserProgressStorage {
```

Y-Pazniak replied:
Fixed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UserFileNotFoundException.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public String getMessage() {
+        return bundle.getString("answers.user_not_found");
```
In my view Exceptions should not own bundle 
```suggestion
        return BundleResourcesServant.getInstance().getString("answers.user_not_found");
```

Y-Pazniak replied:
I couldn't even imagine this construction. Sure, now we can remove bundle and use new construstion

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -1,72 +1,112 @@
 public class UserProgressManager {
-    //private final UserProgress userProgress;
     private final MessagesCreator messagesCreator;
+    private final FileUserProgressStorage fileUserProgressStorage;
```
Please, before using any class think "Can I use abstraction?"
```suggestion
    private final IUserProgressStorage userProgressStorage;
```

Y-Pazniak replied:
OK, thats a great trick. Before creating an object ask myself "is it possible to use abstr.?"
Thank you

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -1,72 +1,112 @@
+            }
+            fileUserProgressStorage.save(userProgress);
+            return messagesCreator.getCountryOrigin(userProgress);
+        } catch (UserFileNotFoundException e) {
```
I think it's better to move catch logic to the router. but this is a controversial decision in any case. Please, do as you feel in this case.

Y-Pazniak replied:
It is as broad as it is long. I don't like both of this ways - they made from our pretty code some kind of spagetti-code.

Y-Pazniak committed:
SendMessage now accepts parameters into constructor.
"D:/" removed into a local variable.
IUserProgressStorage naming was fixed.
String.format was incapsulated into a dedicated  method.
 Redundant inner block if-else in FileUserProgressStorage was deleted.
Saving into the map WAS NOT FIXED.
Reduntant bundle from exceptions was removed.
ReadingUserProgressFromFileException was added.
Final params for methods was added.
Catch logic was removed from UserProgressManager to MessageRouter

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -33,32 +35,33 @@ public SendMessage route(Update update) {
+                default -> answer = getSorryMessage();
+            }
+        } catch (UserFileNotFoundException e) {
+            return new CustomSendMessage(chatID, e.getMessage());
```
Thibk about to translate it here. Exception contains only error code.

Y-Pazniak replied:
Error code was added

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,68 @@
+import java.nio.file.Paths;
+
+public class FileUserProgressStorage implements IUserProgressStorage {
+    private final String databaseEnvVariable = System.getenv("CHATBOT_DATABASE_STORAGE");
```
I think it corresponds to trues more.
```suggestion
    private final String databaseEnvVariable = System.getenv("CHATBOT_FILE_STORAGE");
```

Y-Pazniak replied:
Yes, fixed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public void save(final UserProgress userProgress) {
+        try {
+            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(buildFilePath(userProgress.getChatID())));
```
Please, use [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

Y-Pazniak replied:
Done, thanks for reminding about this construction

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,68 @@
+            ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(buildFilePath(userProgress.getChatID())));
+            oos.writeObject(userProgress);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
It's better to use your custom exception. For example `WritingUserProgressToFileException`

Y-Pazniak replied:
Done

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,68 @@
+        }
+        ObjectInputStream ois;
+        try {
+            ois = new ObjectInputStream(new FileInputStream(buildFilePath(chatID)));
```
Please, use try-with-resources.

Y-Pazniak replied:
Done

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public void createNewUser (final String chatID) {
```
In mu view, just `create` will more correlate with other method `save`, `get`.
```suggestion
    public void create(final String chatID) {
```

Y-Pazniak replied:
Renamed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/MapUserProgressStorage.java:
```diff
@@ -0,0 +1,52 @@
+        private static final MapUserProgressStorage USER_PROGRESS_STORAGE = new MapUserProgressStorage();
+    }
+
+    private Map<String, UserProgress> users;
```
```suggestion
    private final Map<String, UserProgress> users;
```

Y-Pazniak replied:
Fixed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UserFileNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import by.custom.utilcalculator.service.BundleResourcesServant;
+
+public class UserFileNotFoundException extends Exception {
```
I think, it might be helpful if with and other exception contains more info. Like `file path`.

Y-Pazniak replied:
I tried to add more info into exceptions, but I am not sure about success.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/CustomSendMessage.java:
```diff
@@ -0,0 +1,11 @@
+public class CustomSendMessage extends SendMessage {
+
+    public CustomSendMessage(final String chatID, final String answer) {
+        super.setChatId(chatID);
```
maybe there is constructor `super(chatID, answer)`?

Y-Pazniak replied:
Yes, this construction really was there and I could use it without a custom wrapper. Thank you for advice, fixed

Y-Pazniak committed:
1. MapUserProgress doesn't save into file anymore.
2. MessageRouter contains another method for better exceptions handling.
3. UserProgressManager throws exceptions into Router.
4. Exceptions contains their own error codes. Doesn't know about bundle anymore.
5. Abstract UtilsborException was created.
6. Abstract class contains more methods for proper exceptions handling.
7. SendMessage - custom class was deleted, now we use super-class constructor.
8. Comments from PullRequest were fixed.

Y-Pazniak committed:
Postgresql connection was established.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,20 @@
+import org.telegram.telegrambots.meta.api.objects.Message;
+
+public abstract class UtilsborException extends Exception {
+    public static final String USER_FILE_NOT_FOUND = "user_file_not_found";
```
You could move it to enum `UtilsborErrorCode`

Y-Pazniak replied:
Done.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,20 @@
+    public static final String SAVING_INTO_FILE_EXCEPTION = "file_saving_error";
+    public abstract String getExceptionCode();
+
+    public abstract String getFileNameExceptionDetails(final String filename);
```
What if exception is not connected with file. Database exception for example. I think this method is not needed

Y-Pazniak replied:
Removed.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,20 @@
+
+    public abstract String getUserCommandExceptionDetails(final String userCommand);
+
+    public String[] getMessageInfo(final Message message) {
```
Message info is redundant too

Y-Pazniak replied:
Deleted

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,20 @@
+
+    public abstract String getFileNameExceptionDetails(final String filename);
+
+    public abstract String getUserChatIDExceptionDetails(final String chatID);
```
Even this method is not required.

Y-Pazniak replied:
Deleted

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,20 @@
+
+    public abstract String getUserChatIDExceptionDetails(final String chatID);
+
+    public abstract String getUserCommandExceptionDetails(final String userCommand);
```
And this. You can move directly to children class

Y-Pazniak replied:
Deleted

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/MessageRouter.java:
```diff
@@ -33,32 +37,48 @@ public SendMessage route(Update update) {
         String answer;
+        try {
+            answer = route(usersMessage, chatID);
+        } catch (UserFileNotFoundException | ReadingUserProgressFromFileException |
```
```suggestion
        } catch (UtilsborException e)
```

Y-Pazniak replied:
Fixed

Y-Pazniak committed:
Abstract class has two working useful methods.
Abstract class returns error code.
Error classes have no more methods with params.
Abstract class constructor takes 2 params: an error code and the message for printStackTrace();
Each exception class has his own fields for exception details.
Use super() widely (for my own classes and for library classes).
Use abstract Exception where it is possible.

Y-Pazniak committed:
Enums for exceptions were added for

Y-Pazniak committed:
Getters and fields for error classes were added

Y-Pazniak committed:
Getters and fields for error classes were added

Y-Pazniak committed:
Additional constructor for UtilsborException was added.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/controller/Bot.java:
```diff
@@ -22,6 +23,7 @@ public static Bot getBot() {
     //private constructor to avoid wrong bot's creation
     private Bot() {
         messageRouter = MessageRouter.getInstance();
+        initDB();
```
it must be the first

Y-Pazniak replied:
removed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/PostgresUserProgress.java:
```diff
@@ -0,0 +1,22 @@
+public class PostgresUserProgress {
+
+    public void testDatabase() {
+        String jdbcURL = "jdbc:postgresql://localhost:5432/test";
```
Please, take this param from env variables

Y-Pazniak replied:
Done

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/IUserProgressStorage.java:
```diff
@@ -0,0 +1,15 @@
+
+    void create(final String chatID) throws WritingUserProgressIntoFileException;
+
+    String getPath(final String chatID);
```
I suggest this method should not be there, because it's specific for FileStorage

Y-Pazniak replied:
Deleted

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,72 @@
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(filePath))){
+            oos.writeObject(userProgress);
+        } catch (IOException e) {
+            throw new WritingUserProgressIntoFileException(chatID, filePath);
```
Please, add the cause IOException to the constructor as well

Y-Pazniak replied:
Great advice, done

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,72 @@
+        try (ObjectInputStream ois  = new ObjectInputStream(new FileInputStream(filePath))){
+            return (UserProgress) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            throw new ReadingUserProgressFromFileException(chatID, filePath);
```
Please, add the cause Exception to the constructor as well

Y-Pazniak replied:
Done, thanks

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    public void create(final String chatID) throws WritingUserProgressIntoFileException {
```
I think this method is redundant. `save` is enough

Y-Pazniak replied:
Fixed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/MapUserProgressStorage.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class MapUserProgressStorage implements IUserProgressStorage {
```
```suggestion
public class MemoryUserProgressStorage implements IUserProgressStorage {
```

Y-Pazniak replied:
Renamed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/exception/UtilsborException.java:
```diff
@@ -0,0 +1,22 @@
+        this.errorCode = errorCode;
+    }
+
+    public UtilsborException(final String errorCode, final String detailsAboutError) {
```
```suggestion
    public UtilsborException(final String errorCode, final String errorDetails) {
```

Y-Pazniak replied:
Renamed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -1,72 +1,98 @@
 public class UserProgressManager {
-    //private final UserProgress userProgress;
     private final MessagesCreator messagesCreator;
+    private final IUserProgressStorage fileUserProgressStorage;
```
```suggestion
    private final IUserProgressStorage userProgressStorage;
```

Y-Pazniak replied:
Renamed

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -1,72 +1,98 @@
-            case Command.MORE_7_YEARS_AGE -> userProgress.setCarAge(CarAge.MORE_7_YEARS);
-        }
-        return messagesCreator.getCountryOrigin(userProgress);
+    public String processOwnerType(final String command, final String chatID) throws UserFileNotFoundException, ReadingUserProgressFromFileException, WritingUserProgressIntoFileException {
```
```suggestion
    public String processOwnerType(final String command, final String chatID) throws UtilsborException {
```

Y-Pazniak replied:
Done, sorry for this mistake

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/UserProgressManager.java:
```diff
@@ -1,72 +1,98 @@
+                case Command.PHYSICAL_PERSON -> userProgress.setOwnersType(OwnersType.PHYSICAL);
+            }
+            fileUserProgressStorage.save(userProgress);
+            return messagesCreator.getCountryOrigin(userProgress);
```
Are you sure that `getCountryOrigin` should be there?

Y-Pazniak replied:
From this method starts to build string with a user choise. I have added a wrapper-method to imrove this code. It works fine, but naming was really confusing

MaximPilipeiko reviewed (changes requested):
Good. Just few questions

Y-Pazniak committed:
Confusing naming was fixed.
IOException cause was added into custom exceptions' constructor.
Redudant methods was deleted.
Abstractions except realizations were added.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/FileUserProgressStorage.java:
```diff
@@ -0,0 +1,73 @@
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(filePath))){
+            oos.writeObject(userProgress);
+        } catch (IOException e) {
+            throw new WritingUserProgressIntoFileException(chatID, filePath, e.getCause());
```
The cause of `WritingUserProgressIntoFileException` is `IOException` and the cause of IOException is e.getCause(). We must not lose info of IOException.
```suggestion
            throw new WritingUserProgressIntoFileException(chatID, filePath, e);
```

Y-Pazniak replied:
It confuses me, but I think, I got it. 
It helps us to get this kind of message: https://c2n.me/4jXbK1z 
If we send just e.getCause() - we have message in stacktrace without Caused by:

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/domain/MemoryUserProgressStorage.java:
```diff
@@ -0,0 +1,37 @@
+    private final Map<String, UserProgress> users;
+
+    private MemoryUserProgressStorage() {
+        users = new HashMap<>();
```
It's better to use cuncurrent collection here. Just to think. It's not call to action.

Y-Pazniak replied:
What to think about here? We suppose thousands of users - we have to do map cuncurrent.

MaximPilipeiko commented on src/main/java/by/custom/utilcalculator/service/MessagesCreator.java:
```diff
@@ -192,19 +191,23 @@ private String checkForGasolineAuoAge(UserProgress userProgress) {
 
-    private String trimFirstAndLastLetters(String toTrim) {
+    public String getSummaryAnswer(final UserProgress userProgress) {
+        return getCountryOrigin(userProgress);
```
We have to discuss it. For me looks strange that `getSummaryAnswer` returns only `getCountryOrigin`.

Y-Pazniak replied:
Yes, it really looks wrong. But I don't know ho to do it better. 
The logic of building message for user works from getCountryOrigin method. It starts to work here: https://c2n.me/4jXc5Uk - and runs until the message completed. 
Probably, we should rename getCountryOrigin into "getMessageForUser" or something like that? 
I guess, it is more naming problem.

Y-Pazniak committed:
Sending object into exception added.
HashMap has been changed to ConcurrentHashMap.
